src/node_jsrtapi.cc

```diff
@@ -126,51 +127,100 @@ namespace v8impl {
   }
 }  // end of namespace v8impl
 
-void CHAKRA_CALLBACK JsObjectWrapWrapperBeforeCollectCallback(JsRef ref, void* callbackState);
+namespace jsrtimpl {
 
-class ObjectWrapWrapper : public node::ObjectWrap {
-public:
-  ObjectWrapWrapper(napi_value jsObject, void* nativeObj, napi_destruct destructor) {
-    _destructor = destructor;
-    _nativeObj = nativeObj;
+  // Adapter for JSRT external data + finalize callback.
+  class ExternalData {
+  public:
+    ExternalData(void* data, napi_finalize finalize_cb) : _data(data), _cb(finalize_cb) {
+    }
 
-    ObjectWrapWrapper::Wrap(jsObject, this);
+    void* Data() {
+      return _data;
+    }
 
-    JsSetObjectBeforeCollectCallback(reinterpret_cast<JsRef>(jsObject), this, JsObjectWrapWrapperBeforeCollectCallback);
-  }
+    // JsFinalizeCallback
+    static void CALLBACK Finalize(void* callbackState) {
+      ExternalData* externalData = reinterpret_cast<ExternalData*>(callbackState);
+      if (externalData != nullptr) {
+        if (externalData->_cb != nullptr) {
+          externalData->_cb(externalData->_data);
+        }
 
-  void* getValue() {
-    return _nativeObj;
-  }
+        delete externalData;
```

Comment:
This `externalData` pointer is only ever allocated internally. The `_data` member holds the user's data, which may be freed by the provided callback field, `_cb`.